# Wstępne przetwarzanie danych

W tym rozdziale połączone zostaną 

In [1]:
import pandas as pd
import geopandas as gpd

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 2000)
pd.set_option('display.max_rows', 100)

In [3]:
dzielnice_gdf = gpd.read_file("data/dzielnice_lublin.gml", encoding="UTF-8")
lokale_gdf = gpd.read_file("data/lokale_lublin.gml", encoding="UTF-8")

Ramka danych z informacjami o dzielnicach

In [4]:
dzielnice_gdf.head()

,gml_id,lowerCorner,upperCorner,nazwa,geometry
0,dzielnice_granice.1164,5679539.8 8396672.33,5681473.71 8398915.85,Wieniawa,"POLYGON ((8398914.54 5680522.87, 8398914.54 56..."
1,dzielnice_granice.1168,5680335 8395323.055360263,5681595.84 8397528.28,Sławinek,"POLYGON ((8397243.25 5680981.27, 8397304.275 5..."
2,dzielnice_granice.1174,5679808.28 8400476.47,5681940.51 8402927.99,Kalinowszczyzna,"POLYGON ((8401271.162 5680090.935, 8401272.63 ..."
3,dzielnice_granice.1177,5676422.16 8391973.45,5679180.58 8395751.28,Węglin Pd.,"POLYGON ((8395320.29 5678351.25, 8395285.53 56..."
4,dzielnice_granice.1186,5677829.04 8400369.84,5679813.14 8403222.09,Bronowice,"POLYGON ((8403064.64 5678903.88, 8403222.09 56..."


Ramka danych z informacjami o lokalach

In [5]:
lokale_gdf.head()

,gml_id,serwis_rcn,teryt,tran_przestrzen_nazw,tran_lokalny_id_iip,tran_wersja_id,tran_rodzaj_trans,tran_rodzaj_rynku,tran_sprzedajacy,tran_kupujacy,tran_cena_brutto,tran_vat,dok_data,nier_rodzaj,nier_prawo,nier_udzial,nier_pow_gruntu,nier_cena_brutto,nier_vat,lok_id_lokalu,lok_nr_lokalu,lok_funkcja,lok_liczba_izb,lok_nr_kond,lok_pow_uzyt,lok_pow_przyn,lok_cena_brutto,lok_vat,lok_adres,geometry
0,lokale.7652240,None,0609,PL.PZGiK.9349.RCN,5A4905BC-D6A2-40D6-8353-6E94AF99B6AD,NaN,wolnyRynek,NaN,osobaPrawna,osobaFizyczna,370000,NaN,2026-01-26 01:00:00+01,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,0.14,370000,NaN,060911_2.0007.494_BUD.1_LOK,1_LOK,mieszkalna,NaN,NaN,35.8,NaN,NaN,NaN,NaN,POINT (749233.476 386125.694)
1,lokale.7702785,None,0663,PL.PZGiK.4884.RCN,98781cfc-2bb7-4d85-9664-c969a4b2692f,2013-02-21T14:26:22,sprzedazBezprzetargowa,pierwotny,osobaPrawna,osobaFizyczna,20000,NaN,2012-09-27 02:00:00+02,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,NaN,20000,NaN,066301_1.0027.AR_12.34/1.1_BUD.33_LOK,33_LOK,garaz,1,0,396,NaN,20000,NaN,MSC:Lublin;UL:Kryształowa;NR_PORZ:30,POINT (744316.367 378735.781)
2,lokale.7691559,None,0663,PL.PZGiK.4884.RCN,dd4ba0a2-1bf5-4246-913b-643e6abbecad,2013-01-11T11:11:48,wolnyRynek,wtorny,osobaFizyczna,osobaFizyczna,160000,NaN,2012-09-24 02:00:00+02,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,NaN,160000,NaN,066301_1.0019.AR_9.9/8.1_BUD.1_LOK,1_LOK,mieszkalna,3,4,45.84,NaN,160000,NaN,MSC:Lublin;UL:Puchacza;NR_PORZ:8,POINT (750569.449 380204.84)
3,lokale.7682760,None,0663,PL.PZGiK.4884.RCN,33e4defe-db48-4760-8efd-ce127c4b8d75,2013-03-14T09:57:40,wolnyRynek,wtorny,osobaFizyczna,osobaFizyczna,257000,NaN,2012-12-07 01:00:00+01,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,NaN,257000,NaN,066301_1.0006.AR_5.1/4.13_BUD.1_LOK,1_LOK,mieszkalna,4,1,69.6,4.4,257000,NaN,MSC:Lublin;UL:Leszetyckiego;NR_PORZ:6,POINT (747737.827 384676.722)
4,lokale.7693503,None,0663,PL.PZGiK.4884.RCN,d2afa76b-927a-4325-9f0f-74c3c9fbc15a,2013-07-05T14:01:57,sprzedazBezprzetargowa,pierwotny,osobaPrawna,osobaFizyczna,225000,16666.67,2013-04-22 02:00:00+02,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,NaN,225000,16666.67,066301_1.0024.AR_7.12/2.1_BUD.86_LOK,86_LOK,mieszkalna,3,4,49.62,10.79,225000,16666.67,MSC:Lublin;UL:Strzeszewskiego;NR_PORZ:17,POINT (749805.079 384621.036)


## Filtracja lokali na mieszkalne i sprzedane w 2025/2026

Pierwszym krokiem jest wybranie tranzakcji, które zostały dokonane w 2025 lub 2026 roku oraz dotyczyły lokali mieszkalych.

In [6]:
lokale_gdf.shape


(96666, 30)

In [7]:
lokale_gdf["dok_data"] = pd.to_datetime(lokale_gdf["dok_data"], utc=True)

mask = (lokale_gdf["lok_funkcja"] == "mieszkalna") & (
    lokale_gdf["dok_data"].dt.year.isin([2025, 2026]))

lokale_gdf = lokale_gdf[mask]

Rozmiar ramki danych po filtracji:

In [8]:
lokale_gdf.shape

(4249, 30)

Po filtracji pozostało 4249 wiersze, na których będzie przeprowadzana analiza oraz próba predykcji ceny tranzakcyjnej.

## Zmiana crs

Ważnym krokiem w pracy z danymi przestrzennymi jest ujednolicenie używanego crs, tak aby można było wzbogacić dane o mieszkaniach informacją o dzielnicy w której się one znajdują. 

CRS w ramce danych z mieszkaniami:

In [9]:
lokale_gdf.crs

<Projected CRS: EPSG:2180>
Name: ETRF2000-PL / CS92
Axis Info [cartesian]:
- x[north]: Northing (metre)
- y[east]: Easting (metre)
Area of Use:
- name: Poland - onshore and offshore.
- bounds: (14.14, 49.0, 24.15, 55.93)
Coordinate Operation:
- name: Poland CS92
- method: Transverse Mercator
Datum: ETRF2000 Poland
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

CRS ramki danych z dzielnicami:

In [10]:
dzielnice_gdf.crs

<Projected CRS: EPSG:2179>
Name: ETRF2000-PL / CS2000/24
Axis Info [cartesian]:
- x[north]: Northing (metre)
- y[east]: Easting (metre)
Area of Use:
- name: Poland - east of 22°30'E.
- bounds: (22.5, 49.0, 24.15, 54.41)
Coordinate Operation:
- name: Poland CS2000 zone 8
- method: Transverse Mercator
Datum: ETRF2000 Poland
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

CRS są różne zatem CRS dzielnicy został zmieniony na EPSG:2180.

In [11]:
dzielnice_gdf = dzielnice_gdf.to_crs("EPSG:2180")

CRS ramki danych z dzielnicami po zmianie:

In [12]:
dzielnice_gdf.crs

<Projected CRS: EPSG:2180>
Name: ETRF2000-PL / CS92
Axis Info [cartesian]:
- x[north]: Northing (metre)
- y[east]: Easting (metre)
Area of Use:
- name: Poland - onshore and offshore.
- bounds: (14.14, 49.0, 24.15, 55.93)
Coordinate Operation:
- name: Poland CS92
- method: Transverse Mercator
Datum: ETRF2000 Poland
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

## Dodanie informacji o dzielnicach

In [13]:
# Funkcja do dodania kolumny dzielnica
def dodaj_dzielnice_do_lokali(gdf_lokale, gdf_dzielnice, nazwa_kolumny_dzielnica='dzielnica', nazwa_kolumny_nazwa='name'):


    gdf_result = gpd.sjoin(gdf_lokale, gdf_dzielnice[['geometry', nazwa_kolumny_nazwa]], 
                            how='left', predicate='within')
    
    if nazwa_kolumny_nazwa in gdf_result.columns:
        gdf_result[nazwa_kolumny_dzielnica] = gdf_result[nazwa_kolumny_nazwa]
        gdf_result = gdf_result.drop(columns=[nazwa_kolumny_nazwa])
    
    if 'index_right' in gdf_result.columns:
        gdf_result = gdf_result.drop(columns=['index_right'])

    lokale_z_dzielnica = gdf_result[gdf_result[nazwa_kolumny_dzielnica].notna()].shape[0]
    lokale_bez_dzielnica = gdf_result[gdf_result[nazwa_kolumny_dzielnica].isna()].shape[0]
    
    print(f"  Lokale z przypisaną dzielnicą: {lokale_z_dzielnica}")
    print(f"  Lokale bez przypisanej dzielnicy: {lokale_bez_dzielnica}")
    
    
    return gdf_result


Lokale zostały połączone z dzielnicami przy pomocy `spatial join` z predykatem `within`, które na podstawie współrzędnych sprawdza czy mieszkanie znajduje się wewnątrz dzielnicy.

In [14]:
combined_gdf = dodaj_dzielnice_do_lokali(lokale_gdf, dzielnice_gdf, "dzielnica", "nazwa")

  Lokale z przypisaną dzielnicą: 4232
  Lokale bez przypisanej dzielnicy: 17


Udało się połączyć 4232 z ich dzielnicami jednak 17 mieszkań zostało bez dzielnicy. Mieszkania te zostały usunięte

Po tej operacji ramka ma następujące wymiary:

In [15]:
# Usuwanie wierszy bez dzielnicy
combined_gdf = combined_gdf[combined_gdf['dzielnica'].notna()]
combined_gdf.shape


(4232, 31)

Poniżej została przedstawiona ramka danych po całym etapie:

In [16]:
combined_gdf.head()

,gml_id,serwis_rcn,teryt,tran_przestrzen_nazw,tran_lokalny_id_iip,tran_wersja_id,tran_rodzaj_trans,tran_rodzaj_rynku,tran_sprzedajacy,tran_kupujacy,tran_cena_brutto,tran_vat,dok_data,nier_rodzaj,nier_prawo,nier_udzial,nier_pow_gruntu,nier_cena_brutto,nier_vat,lok_id_lokalu,lok_nr_lokalu,lok_funkcja,lok_liczba_izb,lok_nr_kond,lok_pow_uzyt,lok_pow_przyn,lok_cena_brutto,lok_vat,lok_adres,geometry,dzielnica
182,lokale.7761778,None,0663,PL.PZGiK.4884.RCN,489235f9-11ae-4ad4-9e02-c7d5de4e955f,2025-02-17T09:45:21,wolnyRynek,wtorny,osobaFizyczna,osobaPrawna,327000,24222.22,2025-01-28 00:00:00+00:00,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,656,327000,24222.22,066301_1.0019.AR_9.5/6.1_BUD.7_LOK,7_LOK,mieszkalna,3,1,45.1,NaN,NaN,NaN,MSC:Lublin;UL:Sowia;NR_PORZ:1,POINT (750558.696 380348.746),Bronowice
183,lokale.7766269,None,0663,PL.PZGiK.4884.RCN,d861fefd-4d61-4dfe-9428-a1eb9cd289e0,2025-08-13T09:33:01,sprzedazBezprzetargowa,pierwotny,osobaPrawna,osobaFizyczna,401702.81,29755.76,2025-07-25 00:00:00+00:00,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,8787,401702.81,29755.76,066301_1.0004.AR_3.15/7.1_BUD.71_LOK,71_LOK,mieszkalna,2,4,33.59,NaN,NaN,NaN,MSC:Lublin;UL:Mirosława Dereckiego;NR_PORZ:13,POINT (746892.793 385137.992),Czechów Pł.
184,lokale.7765672,None,0663,PL.PZGiK.4884.RCN,898950a5-338e-422c-9863-158d9a6cc814,2025-09-11T09:28:18,sprzedazBezprzetargowa,pierwotny,osobaPrawna,osobaFizyczna,443920,32882.96,2025-06-30 00:00:00+00:00,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,6078,443920,32882.96,066301_1.0016.AR_1.6/4.12_BUD.19_LOK,19_LOK,mieszkalna,2,3,45.35,4.64,NaN,NaN,MSC:Lublin;UL:Garbarska;NR_PORZ:8b,POINT (749493.629 380185.463),Kośminek
185,lokale.7765644,None,0663,PL.PZGiK.4884.RCN,f32e9587-2f96-44cf-b339-e24e58e6332c,2026-01-12T15:06:09,sprzedazBezprzetargowa,pierwotny,osobaPrawna,osobaFizyczna,508446,37662.67,2025-09-16 00:00:00+00:00,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,18888,508446,37662.67,066301_1.0033.AR_9.30/2.6_BUD.88_LOK,88_LOK,mieszkalna,2,7,43.61,3.06,NaN,NaN,MSC:Lublin;UL:Nałęczowska;NR_PORZ:67,POINT (744819.35 381977.039),Sławinek
186,lokale.7767388,None,0663,PL.PZGiK.4884.RCN,e2c5d9a8-55dd-492c-9a0a-8478e02f2b54,2026-02-05T15:26:51,sprzedazBezprzetargowa,pierwotny,osobaPrawna,osobaFizyczna,798000,59111.11,2025-12-08 00:00:00+00:00,nieruchomoscLokalowa,wlasnoscLokaluWrazZPrawemZwiazanym,1/1,3419,798000,59111.11,066301_1.0018.AR_6.1/11.4_BUD.46_LOK,46_LOK,mieszkalna,3,7,83.78,5.3,NaN,NaN,MSC:Lublin;UL:Marii Walciszewskiej;NR_PORZ:1,POINT (748162.926 384241.018),Czechów Pł.


In [17]:
combined_gdf.to_file("data/dane_epsg2180_lublin.gml", driver='GML')